### DISCUSSION TITLE
how to save a string with newlines to a variable?

### DISCUSSION THREAD

**Discussion by danielb2** [fish-shell/fish-shell#10498] [General]
```fish
# this works fine
$ echo '{"response":"\nhello\nand\nhi\n\n"}' | jq .response -r

hello
and
hi

# setting it to a variable loses the newlines
$ set -l foo (echo '{"response":"\nhello\nand\nhi\n\n"}' | jq .response -r); echo $foo
 hello and hi
```

**Comment by faho** [comment]
It does not "lose" the newlines.

It *splits* the variable on them.

Run `set --show foo` and it'll tell you something like (I replaced your command with a similar `printf`):

```
$foo: set in local scope, unexported, with 3 elements
$foo[1]: |hello|
$foo[2]: |and|
$foo[3]: |hi|
```

There are two ways to get the elements as one: `string collect` and `"$()"`:

```fish
set -l foo "$(echo '{"response":"\nhello\nand\nhi\n\n"}' | jq .response -r)"

# or

set -l foo (echo '{"response":"\nhello\nand\nhi\n\n"}' | jq .response -r | string collect)
```

The main difference in the latter is that you can pass more options to control what you want, like `--no-trim-newlines` to not remove the trailing newlines.

This is talked about both in https://fishshell.com/docs/current/language.html#command-substitution and https://fishshell.com/docs/current/fish_for_bash_users.html#command-substitutions.
